src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp

```diff
@@ -177,38 +164,35 @@ void MergeTreeBackgroundExecutor::routine(ItemPtr item)
 }
 
 
-void MergeTreeBackgroundExecutor::schedulerThreadFunction()
+void MergeTreeBackgroundExecutor::threadFunction(size_t number)
 {
+    setThreadName(name.c_str());
+
     while (true)
     {
-        std::unique_lock lock(mutex);
-
-        has_tasks.wait(lock, [this](){ return !pending.empty() || shutdown; });
+        ItemPtr item;
+        {
+            std::unique_lock lock(mutex);
+            has_tasks.wait(lock, [this](){ return !pending.empty() || shutdown; });
 
-        if (shutdown)
-            break;
+            if (number >= threads_count)
+                break;
 
-        ItemPtr item = std::move(pending.front());
-        pending.pop_front();
+            if (shutdown)
+                break;
 
-        /// Execute a piece of task
-        bool res = pool.trySchedule([this, item] () mutable
-        {
-            routine(item);
-            /// When storage shutdowns it will wait until all related background tasks
-            /// are finished, because they may want to interact with its fields
-            /// and this will cause segfault.
-            if (item->is_currently_deleting)
-                item->is_done.set();
-        });
-
-        if (!res)
-        {
-            pending.push_back(item);
-            continue;
+            item = std::move(pending.front());
+            pending.pop_front();
+            active.push_back(item);
         }
 
-        active.push_back(std::move(item));
+        routine(item);
+
+        /// When storage shutdowns it will wait until all related background tasks
+        /// are finished, because they may want to interact with its fields
+        /// and this will cause segfault.
+        if (item->is_currently_deleting)
```

Comment:
Also I think we should signal if `task->execute()` returned false.

Reply:
What do you mean? We will always call set() on Poco::Event object after routine() is executed and `is_curretly_deleting=true`

Reply:
I suppose that there may be a potential race.
We don't have lock here, but but check for item's flag. Suppose that `task->execute()` returned false, but task was not removed from `active` yet. Then, in `removeTasksCorrespondingToStorage`, we copy this task into `tasks_to_wait`. Then, task is removed from `active`, we check that `is_currently_deleting` is false and do not signal event. Then, `removeTasksCorrespondingToStorage` wait for task infinitely.

This actually does not happen because we fill `tasks_to_wait` and set `is_currently_deleting` under the same mutex, and it always happens before removing task from `active`.
It may be not obvious, however. More explicit invariant is to signal every time task is removed from both queues (active and pending).